jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/WorkItemNodeInstance.java

```diff
@@ -127,28 +127,34 @@ public void internalTrigger(final NodeInstance from, String type) {
         if (getNodeInstanceContainer().getNodeInstance(getId()) == null) {
             return;
         }
-        // TODO this should be included for ruleflow only, not for BPEL
-        //        if (!Node.CONNECTION_DEFAULT_TYPE.equals(type)) {
-        //            throw new IllegalArgumentException(
-        //                "A WorkItemNode only accepts default incoming connections!");
-        //        }
+
         WorkItemNode workItemNode = getWorkItemNode();
         createWorkItem(workItemNode);
         if (workItemNode.isWaitForCompletion()) {
             addWorkItemListener();
         }
+
+        WorkItemManager workItemManager = (WorkItemManager) ((ProcessInstance) getProcessInstance()).getKnowledgeRuntime().getWorkItemManager();
         String deploymentId = (String) getProcessInstance().getKnowledgeRuntime().getEnvironment().get(EnvironmentName.DEPLOYMENT_ID);
         workItem.setDeploymentId(deploymentId);
         workItem.setNodeInstanceId(this.getId());
         workItem.setNodeId(getNodeId());
+
+        processWorkItemHandler(() -> workItemManager.internalExecuteWorkItem((org.drools.core.process.instance.WorkItem) workItem));
+
+        if (!workItemNode.isWaitForCompletion()) {
+            triggerCompleted();
+        }
+        this.workItemId = workItem.getId();
+    }
+
+    private void processWorkItemHandler(Runnable handler) {
         if (isInversionOfControl()) {
             ((ProcessInstance) getProcessInstance()).getKnowledgeRuntime()
                                                     .update(((ProcessInstance) getProcessInstance()).getKnowledgeRuntime().getFactHandle(this), this);
         } else {
             try {
-                ((WorkItemManager) ((ProcessInstance) getProcessInstance())
-                                                                           .getKnowledgeRuntime().getWorkItemManager()).internalExecuteWorkItem(
-                                                                                                                                                (org.drools.core.process.instance.WorkItem) workItem);
+                handler.run();
```

Comment:
is this spinning a new thread?

Reply:
nope... just convinient interface.

Reply:
this got me really confused... any other convenient interface we could use that does not bring threading into the picture ... or is this just me...

Reply:
https://docs.oracle.com/javase/8/docs/api/java/util/function/package-summary.html

Not as far as I know. Runnable is marked as Functional interface. I can create a new interface just for this case.